File: data_structures/dll.ts
export {};

class ListNode {
  value: any
  next: (ListNode | null)
  prev: (ListNode | null)

  constructor(value: any) {
    this.value = value;
    this.next = null;
    this.prev = null;
  };
};


class DoublyLinkedList {
  head: (ListNode | null) = null
  tail: (ListNode | null) = null
  length: number = 0

  push(val: any) {
    const node = new ListNode(val);

    if (!this.head) {
      this.head = node;
    } else {
      this.tail!.next = node;
      node.prev = this.tail;
    }
    this.tail = node;
    this.length++;

    return this;
  }

  pop() {
    const popped = this.tail;
    if (!popped) return;

    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = this.tail!.prev;
      this.tail!.next = null;
      popped!.prev = null;
    }

    this.length--;

    return popped;
  }

  shift() {
    const shifted = this.head;
    if (!shifted) return;

    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = this.head!.next;
      this.head!.prev = null;
      shifted.next = null;
    }

    this.length--;

    return shifted;
  }

  unshift(val: any) {
    const node = new ListNode(val);

    if (!this.head) {
      this.tail = node;
    } else {
      this.head.prev = node;
      node.next = this.head;
    }

    this.head = node;
    this.length++;

    return this;
  }

  get(idx: number) {
    if(idx < 0 || idx >= this.length) return;

    let curr: ListNode;

    if ((this.length - idx) > idx ) {
      curr = this.head!;
      for (let i = 0; i < idx; i++) {
        curr = curr!.next!;
      }
    } else {
      curr = this.tail!;
      for (let i = this.length - 1; i > idx; i--) {
        curr = curr!.prev!;
      }
    }

    return curr;
  }

  set(idx:number, val: any) {
    const node = this.get(idx);

    if (!node) return false;

    node.value = val;

    return true;
  }

  insert(idx: number, val: any) {
    const node = new ListNode(val);
    const prevNode = this.get(idx - 1);

    if (idx === 0) this.unshift(val);
    else if (idx === this.length) this.push(val);
    else if (!prevNode) return false;
    else {
      const nextNode = prevNode!.next;
      node.next = nextNode;
      node.prev = prevNode!;
      nextNode!.prev = node;
      prevNode!.next = node;
      this.length++;
    }

    return true;
  }

  remove(idx: number) {
    const currNode = this.get(idx);

    if (idx === 0) return this.shift();
    if (idx === this.length - 1) return this.pop();
    if (!currNode || idx >= this.length) return;

    const prevNode = currNode.prev;
    const nextNode = currNode.next;
    prevNode!.next = nextNode;
    nextNode!.prev = prevNode;

    currNode.prev = null;
    currNode.next = null;

    this.length--;
    return currNode;
  }

  print() {
    let arr = [];
    let curr = this.head;
    while(curr) {
      arr.push(curr.value);
      curr = curr.next;
    }
    console.log(arr);
  }
}

const list = new DoublyLinkedList();
list.push(3);
list.push(10);
list.push(8);
list.unshift(5);
list.unshift(6);
list.unshift(7);
// list.set(5, 4);
// list.insert(0, 4);
// list.insert(6, 4);
// list.insert(7, 4);
// list.insert(-2, 4);
list.print();
list.remove(4);
list.print();
console.log(list.get(3));
console.log(list.get(4));
// list.remove(0);
// list.remove(2);
list.remove(7);
list.remove(-7);
// console.log(list.get(2));

File: data_structures/dll.py
class Node:
    def __init__(self, val):
        self.value = val
        self.next = None
        self.prev = None

    def get_next_val(self):
        if not self.next: return None

        return self.next.value

    def get_prev_val(self):
        if not self.prev: return None

        return self.prev.value


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0


    def push(self, val):
        new_node = Node(val)

        if not self.length:
            self.head = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node

        self.tail = new_node
        self.length += 1

        return self

    def pop(self):
        old_tail = self.tail
        if not old_tail: return

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
            old_tail.prev = None

        self.length -= 1

        return old_tail


    def shift(self):
        old_head = self.head
        if not self.length: return

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
            old_head.next = None

        self.length -= 1

        return old_head


    def unshift(self, val):
        new_node = Node(val)

        if not self.length:
            self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head

        self.head = new_node
        self.length += 1

        return self


    def get(self, idx):
        if idx >= self.length or idx < 0: return

        curr = None

        if (self.length - idx) > idx:
            curr = self.head
            for i in range(idx):
                curr = curr.next
        else:
            curr = self.tail
            for i in range(self.length -1, idx, -1):
                curr = curr.prev

        return curr


    def set(self, idx, val):
        node = self.get(idx)

        if not node: return False

        node.value = val

        return True


    def insert(self, idx, val):
        new_node = Node(val)
        prev_node = self.get(idx - 1)

        if idx == 0: self.unshift(val)
        elif idx == self.length: self.push(val)
        elif not prev_node: return False
        else:
            next_node = prev_node.next

            new_node.next = next_node
            new_node.prev = prev_node
            prev_node.next = new_node
            next_node.prev = new_node

            self.length += 1

        return True


    def remove(self, idx):
        curr_node = self.get(idx)

        if idx == 0: return self.shift()
        if idx == self.length - 1: return self.pop()
        if not curr_node: return

        prev_node = curr_node.prev
        next_node = curr_node.next
        prev_node.next = next_node
        next_node.prev = prev_node

        curr_node.prev = None
        curr_node.next = None

        self.length -= 1

        return curr_node


    def print(self):
        display = []

        curr = self.head
        while curr:
            node = {f"val: {curr.value}, prev: {curr.get_prev_val()}, next: {curr.get_next_val()}"}
            display.append(node)
            curr = curr.next

        print(display)

    def print_list(self):
        display = []

        curr = self.head
        while curr:
            display.append(curr.value)
            curr = curr.next

        print(display)

dll = DoublyLinkedList()
dll.push(7)
dll.push(9)
dll.push(11)
dll.push(12)
dll.push(13)
dll.push(14)
dll.push(15)
dll.print_list()
# dll.print()
# 7 9 11 12
print(dll.get(2).value)
dll.set(5, 20)
dll.print_list()
dll.remove(4)
dll.insert(2, 10)
dll.print_list()

print(dll.insert(12, 10))
print(dll.insert(-2, 10))
print(dll.set(-2, 10))
print(dll.set(12, 10))
print(dll.get(-3))
print(dll.get(13))
print(dll.remove(14))
print(dll.remove(-4))
# print(dll.pop().value)
# print(dll.pop().value)
# dll.print_list()

# print(dll.pop().value)
# print(dll.pop().value)
# print(dll.pop())
# print(dll.pop())
# print(dll.pop())
# dll.print_list()